src/job-cache/job-cache.h

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// the `Cache` class provides the full interface
+// the the underlying complete cache directory.
+// This requires interplay between the file system and
+// the database and must be carefully orchestrated. This
+// class handles all those details and provides a simple
+// interface.
+
+#pragma once
+
+#include <errno.h>
+#include <fcntl.h>
+#include <json/json5.h>
+#include <sqlite3.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/file.h>
+#include <sys/ioctl.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <util/execpath.h>
+#include <util/mkdir_parents.h>
+#include <wcl/filepath.h>
+#include <wcl/trie.h>
+#include <wcl/xoshiro_256.h>
+
+#include <algorithm>
+#include <future>
+#include <iostream>
+#include <map>
+#include <random>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <vector>
+
+#include "bloom.h"
+#include "logging.h"
+
+namespace job_cache {
+
+struct CachedOutputFile {
+  std::string path;
+  Hash256 hash;
+  mode_t mode;
+};
+
+struct JobOutputInfo {
+  std::string stdout_str;
+  std::string stderr_str;
+  int ret_code;
+  double runtime, cputime;
+  uint64_t mem, ibytes, obytes;
+};
+
+struct MatchingJob {
+  int64_t job_id;
+  std::vector<CachedOutputFile> output_files;
+  std::vector<std::string> input_files;
+  std::vector<std::string> input_dirs;
+  JobOutputInfo output_info;
+
+  JAST to_json() const {
+    JAST out(JSON_OBJECT);
+    JAST output_files_json(JSON_ARRAY);
+    JAST input_files_json(JSON_ARRAY);
+    JAST input_dirs_json(JSON_ARRAY);
+
+    // List the output files
+    for (const auto &output_file : output_files) {
+      // TODO: It's a shame we're throwing the hash out here :(
```

Comment:
It would require a lot of work to solve this. The "shame" here is that these files later get hashed to separately by wake. Both on cache hit and when writing to the cache. So it would be nice if the cache could say "I already know the hash, no need to hash it wake" and it would be nice if wake only hashed new output files once rather than twice. Basically we're always hashing one more time than we need to be. But getting those systems to cooperate is not easy at this time.

Reply:
Perfect, can you write some summary of that in the comment